### PR TITLE
chore: Moved tern server to worker

### LINKED ISSUE #5902: Move tern server to web worker
Currently tern server takes 30% of the scripting time on keypress. We would gain that time by moving it to a web worker.
Screenshot 2021-07-15 at 4 44 19 PM

### PR DESCRIPTION
**Pull request by arunvjn**
## Description
Code completion in Appsmith code editors is powered by [tern](https://ternjs.net/doc/manual.html). In the current implementation, tern server runs on the main thread. This server gets queried on every keypress and it gets really expensive even on moderately complex Appsmith apps.

This PR moves tern server to another [web worker](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers), thereby improving the scripting time on the main thread by an average of 25% across all operations.

### TL;DR
Moved tern server into another web worker for performance gains.

Fixes #5902

## Type of change
- Perf improvement

## How Has This Been Tested?
- Manual

### Dev activity
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] PR is being merged under a feature flag

### QA activity:
- [ ] Test plan has been approved by relevant developers
- [ ] Test plan has been peer reviewed by QA
- [ ] Cypress test cases have been added and approved by either SDET or manual QA
- [ ] Organized project review call with relevant stakeholders after Round 1/2 of QA
- [ ] Added Test Plan Approved label after reveiwing all Cypress test

### CODE REVIEW

**Review by Rishabh-Rathod — APPROVED** [review]
Changes look good to me. 🚀

File: app/client/src/sagas/PostEvaluationSagas.ts [hunk 1]
```diff
@@ -337,55 +337,36 @@ export function* postEvalActionDispatcher(actions: Array<AnyReduxAction>) {
 // is accurate
 export function* updateTernDefinitions(
   dataTree: DataTree,
-  updates?: DataTreeDiff[],
+  updates: DataTreeDiff[],
+  isCreateFirstTree: boolean,
```
Comment (Rishabh-Rathod) [inline app/client/src/sagas/PostEvaluationSagas.ts:341]: In `isCreateFirstTree=true` flow updates are sent as `undefined`. Hence, the earlier type was correctly set according to that flow. 
```suggestion
  isCreateFirstTree: boolean,
  updates?: DataTreeDiff[],
```

File: app/client/src/sagas/EvaluationsSaga.ts [hunk 2]
```diff
@@ -216,7 +216,12 @@ function* evaluateTreeSaga(
       isCreateFirstTree,
     );
 
-    yield fork(updateTernDefinitions, updatedDataTree, unEvalUpdates);
+    yield fork(
+      updateTernDefinitions,
+      updatedDataTree,
+      unEvalUpdates,
+      isCreateFirstTree,
+    );
```
Comment (Rishabh-Rathod) [inline app/client/src/sagas/EvaluationsSaga.ts:224]: According to above suggestion.  
```suggestion
    yield fork(
      updateTernDefinitions,
      updatedDataTree,
      isCreateFirstTree,
      unEvalUpdates,
    );
```

**Review by Rishabh-Rathod — COMMENTED** [review]
The changes look good. Just a type fix mentioned above.

File: app/client/src/sagas/PostEvaluationSagas.ts (hunk 1 above)
Comment (arunvjn) [reply app/client/src/sagas/PostEvaluationSagas.ts:341]: The split eval & linting PR actually changed it to []. updates for createFirstTree is now an empty array and not undefined. https://github.com/appsmithorg/appsmith/blob/49a98b5048374d358c72c7b81d2253199dad1f71/app/client/src/workers/Evaluation/evaluation.worker.ts#L226

**Review by Rishabh-Rathod — APPROVED** [review]
Looks good to me.
